bindings/rust/bench/src/harness.rs

```diff
@@ -12,14 +19,145 @@ pub enum Mode {
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ECGroup {
+    SECP256R1,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct CryptoConfig {
+    pub cipher_suite: CipherSuite,
+    pub ec_group: ECGroup,
+}
+
+pub trait TlsBenchHarness: Sized {
+    /// Default harness
+    fn default() -> Result<Self, Box<dyn Error>> {
+        Self::new(&CryptoConfig {
+            cipher_suite: CipherSuite::AES_128_GCM_SHA256,
+            ec_group: ECGroup::SECP256R1,
+        })
+    }
+
     /// Initialize buffers, configs, and connections (pre-handshake)
-    fn new() -> Self;
+    fn new(crypto_config: &CryptoConfig) -> Result<Self, Box<dyn Error>>;
 
     /// Run handshake on initialized connection
-    /// Returns error if handshake already happened
-    fn handshake(&mut self) -> Result<(), &str>;
+    /// Returns error if handshake has already completed
+    fn handshake(&mut self) -> Result<(), Box<dyn Error>>;
 
     /// Checks if handshake is finished for both client and server
     fn handshake_completed(&self) -> bool;
+
+    /// Get negotiated cipher suite
+    ///
+    /// Returns `Err` if handshake has not completed
+    fn get_negotiated_cipher_suite(&self) -> Result<CipherSuite, Box<dyn Error>>;
+
+    /// Get whether or negotiated version is TLS1.3
+    ///
+    /// Retuns `Err` if handshake has not completed
+    fn negotiated_tls13(&self) -> Result<bool, Box<dyn Error>>;
+}
+
+/// Wrapper of two shared buffers to pass as stream
+/// This wrapper `read()`s into one buffer and `write()`s to another
+#[derive(Clone)]
+pub struct ConnectedBuffer {
+    recv: Rc<RefCell<VecDeque<u8>>>,
+    send: Rc<RefCell<VecDeque<u8>>>,
+}
+
+impl ConnectedBuffer {
+    /// Make a new struct with new internal buffers
+    pub fn new() -> Self {
+        ConnectedBuffer {
+            recv: Rc::new(RefCell::new(VecDeque::new())),
+            send: Rc::new(RefCell::new(VecDeque::new())),
+        }
+    }
+    /// Make a new struct that shares internal buffers but swapped, ex.
+    /// `write()` writes to the buffer that the inverse `read()`s from
+    pub fn clone_inverse(&self) -> Self {
+        ConnectedBuffer {
+            recv: Rc::clone(&self.send),
+            send: Rc::clone(&self.recv),
+        }
+    }
+}
+
+impl Read for ConnectedBuffer {
+    fn read(&mut self, dest: &mut [u8]) -> Result<usize, std::io::Error> {
+        match self.recv.borrow_mut().read(dest) {
+            // rustls expects WouldBlock on read of length 0
+            Ok(0) => Err(std::io::Error::new(ErrorKind::WouldBlock, "blocking")),
+            Ok(len) => Ok(len),
+            Err(err) => Err(err),
+        }
+    }
+}
+
+impl Write for ConnectedBuffer {
+    fn write(&mut self, src: &[u8]) -> Result<usize, std::io::Error> {
+        self.send.borrow_mut().write(src)
+    }
+    fn flush(&mut self) -> Result<(), std::io::Error> {
+        Ok(()) // data already available to destination
+    }
+}
+
+#[cfg(test)]
+macro_rules! test_tls_bench_harnesses {
```

Comment:
Again, I probably wouldn't include a macro here if you can get away with a loop. Just for readability.

Reply:
Same response as above: not really a better way.